habitat-lab/habitat/tasks/nav/nav.py

```diff
@@ -1554,7 +1565,10 @@ def _step_rel_waypoint(
                 linear_velocity, angular_velocity, time_step=self._time_step
             )
             xyt = self._agent_state_to_xyt(next_agent_state)
-
+            # Update the xyt of the controller
+            self.w2v_controller.set_goal(
+                xyt_waypoint, start=xyt, relative=True
+            )
```

Comment:
This doesn't make sense. If you keep setting the goal relative to the current location, then the robot will continuously chase a moving target instead of stopping at 0.25 (in the case of move_forward).
What happens if you don't add this?

Reply:
If I do not add this, the robot just move spirally, meaning that it outputs linear and angular velocity at the same time

Reply:
Is this an expected thing the robot moves spirally?

Reply:
Another point is that do we really update the robot local transformation matrix? meaning that if the robot moves at `_t=0`, then the transformation at `_t=0.1` cannot use the previous transformation?

Reply:
The spiral movement is not expected.
The reason that it generates angular velocity is that the state feedback (next_agent_state) isn't interpreted to be moving forwards. So the controller tries to rotate and compensate.
You circumvented the problem by telling the robot to always move forward, but you lose feedback information and thus the controller is now just blindly moving forwards instead of trying to stabilize to the waypoint. What we should be doing is to find out and correct why the state feedback is wrong or being interpreted wrong.

Reply:
> Another point is that do we really update the robot local transformation matrix? meaning that if the robot moves at `_t=0`, then the transformation at `_t=0.1` cannot use the previous transformation?

What do you mean by robot local transformation matrix?

Reply:
`xyt_err = xyt_global_to_base(self.xyt_goal, self.xyt_loc)` It seems xyt_err does not output correct values within _t for-loop: 
_t: 0.0 [ 0.50000078 -0.03000045  0.        ]
_t: 0.1 [ 0.50143064 -0.0465724   0.02681698]
_t: 0.2 [ 0.50370458 -0.0598166   0.06015406]
Maybe `xyt_global_to_base()` is doing something strange?

Reply:
wait, but you do update `update_pose_feedback()` each time, which I think it is correct

Reply:
> 

Yes exactly

Reply:
> 

Are you printing out values of `xyt_err`?
If so at _t: 0.0 the error should be [0.25, 0, 0].

Reply:
Yes, I am printing out `xyt_err`, and it should be [0.5, 0, 0]? why  [0.25, 0, 0]?

Reply:
What command are you giving the robot? Are you not calling `move_forward`?

Reply:
This is what I got from running move_forward on the latest branch:
```
======================
action={'action': 'move_forward_waypoint', 'action_args': None}
===========[controller integration]===========
hab_goal=[0.25 0.   0.  ], agent_goal=[ 14.89055808 -17.44277298   4.03323325]
t=0.0
        hab_xyt=[ 15.04759216 -17.24824715   4.03323325], agent_xyt=[ 15.04759216 -17.24824715   4.03323325]
        agent_xyt_err=[ 2.50000000e-01 -1.16573418e-15  0.00000000e+00]
        vel_cmd=[0.3, -7.480323551899881e-08]
t=0.1
        hab_xyt=[ 15.04759216 -17.24824715   4.03323325], agent_xyt=[ 15.04759216 -17.24824715   4.03323325]
        agent_xyt_err=[ 2.50000000e-01 -1.16573418e-15  0.00000000e+00]
        vel_cmd=[0.3, -7.480323551899881e-08]
t=0.2
        hab_xyt=[ 15.04759216 -17.24824715   4.03323325], agent_xyt=[ 15.04759216 -17.24824715   4.03323325]
        agent_xyt_err=[ 2.50000000e-01 -1.16573418e-15  0.00000000e+00]
        vel_cmd=[0.3, -7.480323551899881e-08]
```
As you can see xyt_err is always [0.25, 0, 0] since you are setting the goal every timestep.

Reply:
Also the robot is rotating like 90 degrees every time we call `turn_left` or `turn_right`